blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md

```diff
@@ -0,0 +1,259 @@
+callee.
+Existential types capture this.
+
+We can translate functions in the source of type `A -> B` to
+pairs of an environment and a code pointer, but now making the
+environment type existentially quantified:
+
+```
+∃ Γ. Γ × (Γ × A -> B).
+```
+Then the syntax of existential types ensure that all any consumer can
+do with the `env : Γ` in the pair is pass it to the code
+pointer with an `A` argument.
+
+How do we prove that this is correct? And what does correct even mean?
+We'll focus on a property called *full abstraction* which says that if
+two programs are equal in the source language, then their translations
+are equal.
+Here, equal in the source language will just mean \\(\beta,\eta \\)
+equivalence, so things like as above:
+
+```
+let x = 3 in λ y. x + y ≡ λ y. 3 + y
+```
+
+To prove this we'll show that in a language with existential types the
+types `∃ Γ. Γ × (Γ × A -> B)` and `A \to B` are isomorphic.
+The usual proof is by parametricity, instead we'll use a closely
+related category-theoretic argument: the coYoneda lemma.
+
+## The CoYoneda Lemma
+
+The coYoneda lemma is a generalization of the equivalence described
+above.
+I'll start with the ordinary version which uses *coends* and
+*presheaves*.
+
+The coYoneda lemma says that for any category \\( C \\), *presheaf*,
+i.e. a functor \\( Q : C^{op} \to \Set \\), and object \\(A \in C \\),
+\\(Q(A) \\) is isomorphic to:
+\\[ \exists B. (A \to B) \times Q(B) \\]
+usually coends use an integral sign but I think \\(\exists \\) is a
+much more evocative notation that is used sometimes in logicy circles.
+
+A coend is a construction that is very similar to the parametric
+existential quantifier.
+I don't think I've ever seen this written down explicitly, but it's
+basically what you get when you take the parametricity condition for
+the existential type but restrict your reasoning to only use
+functional relations.
+
+A presheaf is just a functor \\( Q : C^{op} \to \Set \\).  One nice
+way to think about it is that \\( Q \\) is a sort of "generalized
+object" of \\( C \\), in that it has a notion of map from object \\( B
+\in C\\) to \\( Q \\): we think of \\(Q(B) \\) as "\\( B \to Q
+\\)". The fact that \\( Q \\) is a functor means we can compose a
+"map" "\\( q : B \to Q \\)" (really an element \\(q \in Q(B)\\)) with
+real maps \\( f : A \to B \\) in \\( C\\) to get a map \\( q \circ f :
+A \to Q \\) and this behaves as expected with respect to identity and
+composition.  Then the Yoneda lemma says that any object is a
+"generalized object" by taking the real maps into it and that this
+interpretation doesn't add any new maps between the original objects.
+
+So the informal explanation of the coyoneda lemma is that if we have a
+\\( \exists B. (A \to B) \times Q(B) \\) then we can get a \\( f : A
+\to B \\) and a "\\(q : B \to Q\\)" and we can't inspect \\(B \\) in
+any way, then all we can really do is to compose the two maps.
```

Comment:
"then all ...." makes this sentence hard to parse

Maybe say ".... if we have a `\exists` .... can get a `f` ... and a `q`. Furthermore we can compose the 2 maps but we cannot inspect `B` in any way"